File: tests/algorithms/transpiler/FS/other/lfu_cache.fs
// Generated 2025-08-22 13:05 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _arrset (arr:'a array) (i:int) (v:'a) : 'a array =
    let mutable a = arr
    if i >= a.Length then
        let na = Array.zeroCreate<'a> (i + 1)
        Array.blit a 0 na 0 a.Length
        a <- na
    a.[i] <- v
    a
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
type Entry = {
    mutable _key: int
    mutable ``val``: int
    mutable _freq: int
    mutable _order: int
}
type LFUCache = {
    mutable _entries: Entry array
    mutable _capacity: int
    mutable _hits: int
    mutable _miss: int
    mutable _tick: int
}
type GetResult = {
    mutable _cache: LFUCache
    mutable _value: int
    mutable _ok: bool
}
let rec lfu_new (cap: int) =
    let mutable __ret : LFUCache = Unchecked.defaultof<LFUCache>
    let mutable cap = cap
    try
        __ret <- { _entries = Array.empty<Entry>; _capacity = cap; _hits = 0; _miss = 0; _tick = 0 }
        raise Return
        __ret
    with
        | Return -> __ret
and find_entry (_entries: Entry array) (_key: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable _entries = _entries
    let mutable _key = _key
    try
        let mutable i: int = 0
        while i < (Seq.length (_entries)) do
            let e: Entry = _idx _entries (int i)
            if (e._key) = _key then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- 0 - 1
        raise Return
        __ret
    with
        | Return -> __ret
and lfu_get (_cache: LFUCache) (_key: int) =
    let mutable __ret : GetResult = Unchecked.defaultof<GetResult>
    let mutable _cache = _cache
    let mutable _key = _key
    try
        let idx: int = find_entry (_cache._entries) (_key)
        if idx = (0 - 1) then
            let new_cache: LFUCache = { _entries = _cache._entries; _capacity = _cache._capacity; _hits = _cache._hits; _miss = (_cache._miss) + 1; _tick = _cache._tick }
            __ret <- { _cache = new_cache; _value = 0; _ok = false }
            raise Return
        let mutable _entries: Entry array = _cache._entries
        let mutable e: Entry = _idx _entries (int idx)
        e._freq <- (e._freq) + 1
        let new_tick: int = (_cache._tick) + 1
        e._order <- new_tick
        _entries.[idx] <- e
        let new_cache: LFUCache = { _entries = _entries; _capacity = _cache._capacity; _hits = (_cache._hits) + 1; _miss = _cache._miss; _tick = new_tick }
        __ret <- { _cache = new_cache; _value = e.``val``; _ok = true }
        raise Return
        __ret
    with
        | Return -> __ret
and remove_lfu (_entries: Entry array) =
    let mutable __ret : Entry array = Unchecked.defaultof<Entry array>
    let mutable _entries = _entries
    try
        if (Seq.length (_entries)) = 0 then
            __ret <- _entries
            raise Return
        let mutable min_idx: int = 0
        let mutable i: int = 1
        while i < (Seq.length (_entries)) do
            let e: Entry = _idx _entries (int i)
            let m: Entry = _idx _entries (int min_idx)
            if ((e._freq) < (m._freq)) || (((e._freq) = (m._freq)) && ((e._order) < (m._order))) then
                min_idx <- i
            i <- i + 1
        let mutable res: Entry array = Array.empty<Entry>
        let mutable j: int = 0
        while j < (Seq.length (_entries)) do
            if j <> min_idx then
                res <- Array.append res [|(_idx _entries (int j))|]
            j <- j + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and lfu_put (_cache: LFUCache) (_key: int) (_value: int) =
    let mutable __ret : LFUCache = Unchecked.defaultof<LFUCache>
    let mutable _cache = _cache
    let mutable _key = _key
    let mutable _value = _value
    try
        let mutable _entries: Entry array = _cache._entries
        let idx: int = find_entry (_entries) (_key)
        if idx <> (0 - 1) then
            let mutable e: Entry = _idx _entries (int idx)
            e.``val`` <- _value
            e._freq <- (e._freq) + 1
            let new_tick: int = (_cache._tick) + 1
            e._order <- new_tick
            _entries.[idx] <- e
            __ret <- { _entries = _entries; _capacity = _cache._capacity; _hits = _cache._hits; _miss = _cache._miss; _tick = new_tick }
            raise Return
        if (Seq.length (_entries)) >= (_cache._capacity) then
            _entries <- remove_lfu (_entries)
        let new_tick: int = (_cache._tick) + 1
        let new_entry: Entry = { _key = _key; ``val`` = _value; _freq = 1; _order = new_tick }
        _entries <- Array.append _entries [|new_entry|]
        __ret <- { _entries = _entries; _capacity = _cache._capacity; _hits = _cache._hits; _miss = _cache._miss; _tick = new_tick }
        raise Return
        __ret
    with
        | Return -> __ret
and cache_info (_cache: LFUCache) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable _cache = _cache
    try
        __ret <- ((((((("CacheInfo(hits=" + (_str (_cache._hits))) + ", misses=") + (_str (_cache._miss))) + ", capacity=") + (_str (_cache._capacity))) + ", current_size=") + (_str (Seq.length (_cache._entries)))) + ")"
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let mutable _cache: LFUCache = lfu_new (2)
        _cache <- lfu_put (_cache) (1) (1)
        _cache <- lfu_put (_cache) (2) (2)
        let mutable r: GetResult = lfu_get (_cache) (1)
        _cache <- r._cache
        if r._ok then
            ignore (printfn "%s" (_str (r._value)))
        else
            ignore (printfn "%s" ("None"))
        _cache <- lfu_put (_cache) (3) (3)
        r <- lfu_get (_cache) (2)
        _cache <- r._cache
        if r._ok then
            ignore (printfn "%s" (_str (r._value)))
        else
            ignore (printfn "%s" ("None"))
        _cache <- lfu_put (_cache) (4) (4)
        r <- lfu_get (_cache) (1)
        _cache <- r._cache
        if r._ok then
            ignore (printfn "%s" (_str (r._value)))
        else
            ignore (printfn "%s" ("None"))
        r <- lfu_get (_cache) (3)
        _cache <- r._cache
        if r._ok then
            ignore (printfn "%s" (_str (r._value)))
        else
            ignore (printfn "%s" ("None"))
        r <- lfu_get (_cache) (4)
        _cache <- r._cache
        if r._ok then
            ignore (printfn "%s" (_str (r._value)))
        else
            ignore (printfn "%s" ("None"))
        ignore (printfn "%s" (cache_info (_cache)))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
ignore (main())

File: tests/algorithms/x/PHP/other/lfu_cache.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _len($x) {
    if ($x === null) { return 0; }
    if (is_array($x)) { return count($x); }
    if (is_string($x)) { return strlen($x); }
    return strlen(strval($x));
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function lfu_new($cap) {
  return ['capacity' => $cap, 'entries' => [], 'hits' => 0, 'miss' => 0, 'tick' => 0];
};
  function find_entry($entries, $key) {
  $i = 0;
  while ($i < count($entries)) {
  $e = $entries[$i];
  if ($e['key'] == $key) {
  return $i;
}
  $i = $i + 1;
};
  return 0 - 1;
};
  function lfu_get($cache, $key) {
  $idx = find_entry($cache['entries'], $key);
  if ($idx == 0 - 1) {
  $new_cache = ['capacity' => $cache['capacity'], 'entries' => $cache['entries'], 'hits' => $cache['hits'], 'miss' => $cache['miss'] + 1, 'tick' => $cache['tick']];
  return ['cache' => $new_cache, 'ok' => false, 'value' => 0];
}
  $entries = $cache['entries'];
  $e = $entries[$idx];
  $e['freq'] = $e['freq'] + 1;
  $new_tick = $cache['tick'] + 1;
  $e['order'] = $new_tick;
  $entries[$idx] = $e;
  $new_cache = ['capacity' => $cache['capacity'], 'entries' => $entries, 'hits' => $cache['hits'] + 1, 'miss' => $cache['miss'], 'tick' => $new_tick];
  return ['cache' => $new_cache, 'ok' => true, 'value' => $e['val']];
};
  function remove_lfu($entries) {
  if (count($entries) == 0) {
  return $entries;
}
  $min_idx = 0;
  $i = 1;
  while ($i < count($entries)) {
  $e = $entries[$i];
  $m = $entries[$min_idx];
  if ($e['freq'] < $m['freq'] || ($e['freq'] == $m['freq'] && $e['order'] < $m['order'])) {
  $min_idx = $i;
}
  $i = $i + 1;
};
  $res = [];
  $j = 0;
  while ($j < count($entries)) {
  if ($j != $min_idx) {
  $res = _append($res, $entries[$j]);
}
  $j = $j + 1;
};
  return $res;
};
  function lfu_put($cache, $key, $value) {
  $entries = $cache['entries'];
  $idx = find_entry($entries, $key);
  if ($idx != 0 - 1) {
  $e = $entries[$idx];
  $e['val'] = $value;
  $e['freq'] = $e['freq'] + 1;
  $new_tick = $cache['tick'] + 1;
  $e['order'] = $new_tick;
  $entries[$idx] = $e;
  return ['capacity' => $cache['capacity'], 'entries' => $entries, 'hits' => $cache['hits'], 'miss' => $cache['miss'], 'tick' => $new_tick];
}
  if (count($entries) >= $cache['capacity']) {
  $entries = remove_lfu($entries);
}
  $new_tick = $cache['tick'] + 1;
  $new_entry = ['freq' => 1, 'key' => $key, 'order' => $new_tick, 'val' => $value];
  $entries = _append($entries, $new_entry);
  return ['capacity' => $cache['capacity'], 'entries' => $entries, 'hits' => $cache['hits'], 'miss' => $cache['miss'], 'tick' => $new_tick];
};
  function cache_info($cache) {
  return 'CacheInfo(hits=' . _str($cache['hits']) . ', misses=' . _str($cache['miss']) . ', capacity=' . _str($cache['capacity']) . ', current_size=' . _str(_len($cache['entries'])) . ')';
};
  function main() {
  $cache = lfu_new(2);
  $cache = lfu_put($cache, 1, 1);
  $cache = lfu_put($cache, 2, 2);
  $r = lfu_get($cache, 1);
  $cache = $r['cache'];
  if ($r['ok']) {
  echo rtrim(_str($r['value'])), PHP_EOL;
} else {
  echo rtrim('None'), PHP_EOL;
}
  $cache = lfu_put($cache, 3, 3);
  $r = lfu_get($cache, 2);
  $cache = $r['cache'];
  if ($r['ok']) {
  echo rtrim(_str($r['value'])), PHP_EOL;
} else {
  echo rtrim('None'), PHP_EOL;
}
  $cache = lfu_put($cache, 4, 4);
  $r = lfu_get($cache, 1);
  $cache = $r['cache'];
  if ($r['ok']) {
  echo rtrim(_str($r['value'])), PHP_EOL;
} else {
  echo rtrim('None'), PHP_EOL;
}
  $r = lfu_get($cache, 3);
  $cache = $r['cache'];
  if ($r['ok']) {
  echo rtrim(_str($r['value'])), PHP_EOL;
} else {
  echo rtrim('None'), PHP_EOL;
}
  $r = lfu_get($cache, 4);
  $cache = $r['cache'];
  if ($r['ok']) {
  echo rtrim(_str($r['value'])), PHP_EOL;
} else {
  echo rtrim('None'), PHP_EOL;
}
  echo rtrim(cache_info($cache)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
